rest/counted_request_reader.go

```diff
@@ -0,0 +1,42 @@
+// Copyright 2023-Present Couchbase, Inc.
+//
+// Use of this software is governed by the Business Source License included
+// in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+// in that file, in accordance with the Business Source License, use of this
+// software will be governed by the Apache License, Version 2.0, included in
+// the file licenses/APL2.txt.
+
+package rest
+
+import (
+	"io"
+)
+
+type CountedRequestReader struct {
+	numBytes int64
+	reader   io.Reader
+}
+
+// NewReaderCounter returns a new CountedRequestReader storing the io.Reader and bytes read off the reader
+func NewReaderCounter(reader io.Reader) *CountedRequestReader {
+	return &CountedRequestReader{
+		reader: reader,
+	}
+}
+
+// Read overrides the Read method from io package that will add the number of bytes read to the CountedRequestReader struct for
+// retrieval for the stat on the database
+func (c *CountedRequestReader) Read(buf []byte) (int, error) {
+	numBytesRead, err := c.reader.Read(buf)
+
+	c.numBytes += int64(numBytesRead)
+	return numBytesRead, err
+}
+
+// LoadCount will load the number of bytes read for the request from the CountedRequestReader struct
+func (c *CountedRequestReader) LoadCount() int64 {
+	return c.numBytes
+}
```

Comment:
I'd probably call this GetBodyBytes() and read this if we are going to need to read this - this avoids having to read bytes in this function later:

```
			// if auth fails we still need to record bytes read over the rest api for the stat, to do this we need to read
			// the body to populate the bytes count on the CountedRequestReader struct
			_, _ = h.readBody()
			dbContext.DbStats.DatabaseStats.PublicRestBytesRead.Add(h.requestBody.LoadCount())
```

I don't think there's another place we need to read it but this moves code out a really complicated function, and in the case of this call where it does hit an invalid case, it's actually going to be checking the `numBytes` twice anyway, so we might as well do a Just In Time read if we need.

Then we read the bytes from a message if we need to (public db scoped) but not in the other cases that I mentioned.